workflow/workflow_test.go

```diff
@@ -1213,6 +1213,124 @@ func TestInputCancelledStepWorkflow(t *testing.T) {
 	assert.Contains(t, err.Error(), "cannot construct any output")
 }
 
+var inputDisabledStepWorkflow = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        step_enabled:
+          type:
+            type_id: bool
+steps:
+  simple_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    deploy:
+      deployer_name: "test-impl"
+      # Disabling steps waits for the value to be resolved before executing the step.
+      # So if it's working correctly, this should not cause any race conditions.
+      deploy_time: 0 # ms
+    input:
+      wait_time_ms: 20
+    enabled: !expr $.input.step_enabled
+outputs:
+  success:
+    simple_wait_output: !expr $.steps.simple_wait.outputs.success
+`
+
+func TestInputDisabledStepWorkflow(t *testing.T) {
+	// Run a workflow with one step that has its enablement state
+	// set by the input. The output only passes on success, so
+	// this test case also tests that the failure case doesn't
+	// lead to a deadlock.
+	preparedWorkflow := assert.NoErrorR[workflow.ExecutableWorkflow](t)(
+		getTestImplPreparedWorkflow(t, inputDisabledStepWorkflow),
+	)
+	// The workflow should pass with it enabled
+	outputID, _, err := preparedWorkflow.Execute(context.Background(), map[string]any{
+		"step_enabled": true,
+	})
+	assert.NoError(t, err)
+	assert.Equals(t, outputID, "success")
+	// The workflow should fail with it disabled
```

Comment:
Because the workflow output depends on `outputs.success` which will not be resolved (produced) if the step is disabled?  (But, yeah, it would be good if the block comment at the top of the function pointed that out.)